### DISCUSSION TITLE
How to write data variables to a file

### DISCUSSION THREAD

**Discussion by halloleo** [twpayne/chezmoi#3929] [Q&A] [answered]
In my `.chezmoi.toml.tmpl` I set some data variables with a logic like this:

```
{{- if eq .chezmoi.hostname "work-name" }}
sysname = "abc"
sysgroup = "data"
sysarea = "work"
{{- else if eq .chezmoi.hostname "home-name" }}
sysname = "def"
sysgroup = ""
sysarea = "home"
{{- else if eq .chezmoi.hostname ...}}
...
{{- end }}
```

These variables I would like to use in my shell as well. So I thought of writing them from this template to a file `.machine` that my .bash_profile can source. (I don't want to make `.bash_profile` itself a template.) On my work computer for example this file should be written like this:

```
$SYSNAME = "abc"
$SYSGROUP = "data"
$SYSAREA = "work"
```

How can I write to a file in the `.chezmoi.toml.tmpl` template?

**Comment by twpayne** [comment, marked as answer]
> How can I write to a file in the `.chezmoi.toml.tmpl` template?

You can't, but you can make `.machine` a template.

**Reply by halloleo** [reply]
Ah, I see. 

Cool, it works: With `.machine.tmpl` I can write a file `.machine` which contains the particular data variables as shell variables. Excellent.

Is there a way in the template language to check whether a variable exists? (Because when I use a variable in `.machine.tmpl` I have not define in `chezmoi.toml` I get an "_map has no entry for key_" error.)

Or can I list all existing variables in a template?

**Reply by bradenhilton** [reply]
@halloleo Have you tried using the [`dig`](http://masterminds.github.io/sprig/dicts.html#dig) function? You could also try the [Missing keys](https://www.chezmoi.io/reference/templates/directives/#missing-keys) directive.

**Comment by Lockszmith-GH** [comment]
The way I solved this in my envrionment, is by having a function/or script in path named: `cz-import-data`
It relies on `jq` and `sed`

Assuming my `chezmoi` config is as follows:

```console
> find ~/.config/chezmoi -name 'chezmoi.*' -print -exec cat '{}' \;
.../.config/chezmoi/chezmoi.toml
[data]
sysname = "abc"
sysgroup = "data"
sysarea = "work"
```

Defining a PREFIX (in the code below the default is `CZ_DATA_` and make sure the names are uppercase via the UPPERCASE env variable.
The output will look like:

```console
$ import-cz-data
CZ_DATA_SYSAREA="work"
CZ_DATA_SYSGROUP="data"
CZ_DATA_SYSNAME="abc"
```

The idea is that running the following commands allows you to evaluate/source those into environment variables:

```console
. <( import-cz-data )

❯ echo $CZ_DATA_SYSAREA
work
```

This is the code of `cz-import-data`

```bash
cz-import-data() {
    # Use this with eval: eval $(cz-import-data)
    local UPPERCASE='|ascii_upcase'
    local PREFIX="CZ_DATA_"
    [[ "$1" == "promptString" ]] && unset UPPERCASE PREFIX
    local RESULTS=$(
        chezmoi data --format json \
            | jq -r ".chezmoi.config.data" \
            | jq -r "
                to_entries[] |
                    \"$PREFIX\(.key${UPPERCASE})=\\\"\(.value)\\\"\"
            "
        )
    if [[ "$1" == "promptString" ]] ; then
        printf "$RESULTS" | tr '\n' ',' | sed 's/"//g; s/^\(.*\),$/"\1/g;'
    else
        echo "$RESULTS"
    fi  
}
```

**Reply by Lockszmith-GH** [reply]
Oh, I completely forgot, the 'promptString' stuff.

With a `.chezmoi.toml.tmpl` like this:

```ini
{{- $sysname  := promptStringOnce . "sysname"  "System name"  -}}
{{- $sysgroup := promptStringOnce . "sysgroup" "System group" -}}
{{- $sysarea  := promptStringOnce . "sysarea"  "System area"  -}}

[data]
{{- if ($sysname) }}
    sysname = {{ $sysname | quote }}
{{- end -}}
{{- if ($sysgroup) }}
    sysgroup = {{ $sysgroup | quote }}
{{- end -}}
{{- if ($sysarea) }}
    sysarea = {{ $sysarea | quote }}
{{- end }}
```

When initializing it prompts for the values, and to re-initialize you can run the following command:

```console
$ chezmoi init --promptString $(cz-import-data promptString) [--prompt]
System name? one
System group? two
System area? three

$ cz-import-data
CZ_DATA_SYSAREA="three"
CZ_DATA_SYSGROUP="two"
CZ_DATA_SYSNAME="one"
```

The optional `--prompt` will force prompting for all. Without it, it will only prompt for missing data.

Hope this becomes useful to someone.

**Reply by halloleo** [reply]
Thanks for this. Very sophisticated. I keep it in mind as an option
